### PR TITLE
feat: add Carmichael generator

### PR DESCRIPTION
**Pull request by imengus**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: math/carmichael.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{ 
+    tests();
+    return 0;
+}
```
Comment (realstealthninja) [inline math/carmichael.cpp:116]:
```suggestion
}

```

File: math/carmichael.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    mpz_clear(n);
+    mpz_clear(N);
+    mpz_clear(output);
+}
+
+/**
+ * @brief Self-test implementations to test `isCarmichael`
+ * @returns `void`
+ */
+
```
Comment (realstealthninja) [inline math/carmichael.cpp:90]:
```suggestion
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by realstealthninja** [comment]
Thank you so much for contributing! :heart:

File: math/carmichael.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * for all `a` that is coprime with `n`. The smallest example is 561.
+ * Find out more on: https://en.wikipedia.org/wiki/Carmichael_number
+ * 
+ * -lgmp flag may be necessary, e.g.
+ *         gcc -o carmichael carmichael.cpp -lgmp ; ./carmichael
+ */ 
+#include <cassert>    /// for assert
+#include <gmp.h>      /// for std::mpz_*
+#include <cstdio>     /// for std::printf
+using namespace std;
```
[the remark this answers is not in the document]
Comment (imengus) [reply math/carmichael.cpp:18]: I think I have enough functions that prefixing each with std:: would decrease the readability. Are you suggesting that I replace that line with `using std::name` only?

[the remark this answers is not in the document]
Comment (imengus) [reply math/carmichael.cpp:18]: Sorry to admit it but this is my first C/C++ program! 

I misunderstood initially, but trying to prefix each function with std I had the following kind of error for each function: `namespace "std" has no member "__gmpz_clear". I had written std::mpz_* as part of the imports without knowing this error would arise.

When I simply omit `using namespace std;` without changing the code (so without prefixes) the code runs normally. The online examples of gmp code I saw also didn't include this line, so would prefixing still be necessary?

[the remark this answers is not in the document]
Comment (realstealthninja) [reply math/carmichael.cpp:18]: its not necessary.  It is necessary for any function from the standard library i.e cout.
in your case if it works without prefixing then its fine! just omit using namespace std;
its bad practice to do using namespace std hence my initial comment.

File: math/carmichael.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
```
Comment (realstealthninja) [inline math/carmichael.cpp:18]:
```suggestion
```

File: math/carmichael.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    mpz_inits(output, NULL);
+
+    mpz_init_set_ui(n, 561);
+    assert (isCarmichael(n, a, output) == true);
+
+    mpz_set_ui(n, 334153);
+    assert(isCarmichael(n, a, output) == true);
+    
+    mpz_set_ui(n, 1009);
+    assert(isCarmichael(n, a, output) == false);
```
Comment (realstealthninja) [inline math/carmichael.cpp:103]:
```suggestion
    assert(isCarmichael(n, a, output) == false);
    
    std::cout << "All tests have passed successfully";
```

File: math/carmichael.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{ 
+    tests();
```
Comment (realstealthninja) [inline math/carmichael.cpp:113]:
```suggestion
    tests(); // run self-implemented tests
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Pull request closed without merging**